src/tools/spin/spinquic.cpp

```diff
@@ -1054,6 +1057,89 @@ CXPLAT_THREAD_CALLBACK(RunThread, Context)
     CXPLAT_THREAD_RETURN(0);
 }
 
+
+int start(void* Context) {
+    CxPlatSystemLoad();
+    CxPlatInitialize();
+
+    //
+    // Initial MsQuicOpen2 and initialization.
+    //
+    const QUIC_API_TABLE* TempMsQuic = nullptr;
+    ASSERT_ON_FAILURE(MsQuicOpen2(&TempMsQuic));
+    CxPlatCopyMemory(&MsQuic, TempMsQuic, sizeof(MsQuic));
+
+    if (Settings.AllocFailDenominator > 0) {
+        if (QUIC_FAILED(
+            MsQuic.SetParam(
+                nullptr,
+                QUIC_PARAM_GLOBAL_ALLOC_FAIL_DENOMINATOR,
+                sizeof(Settings.AllocFailDenominator),
+                &Settings.AllocFailDenominator))) {
+            printf("Setting Allocation Failure Denominator failed.\n");
+        }
+    }
+
+    if (Settings.LossPercent != 0) {
+        QUIC_TEST_DATAPATH_HOOKS* Value = &DataPathHooks;
+        if (QUIC_FAILED(
+            MsQuic.SetParam(
+                nullptr,
+                QUIC_PARAM_GLOBAL_TEST_DATAPATH_HOOKS,
+                sizeof(Value),
+                &Value))) {
+            printf("Setting Datapath hooks failed.\n");
+        }
+    }
+
+    MsQuicClose(TempMsQuic);
+
+    Settings.RunTimeMs = Settings.RunTimeMs / Settings.RepeatCount;
+    for (uint32_t i = 0; i < Settings.RepeatCount; i++) {
+
+        CXPLAT_THREAD_CONFIG Config = {
+            0, 0, "spin_run", RunThread, Context
+        };
+        CXPLAT_THREAD Threads[4];
+        const uint32_t Count = (uint32_t)(rand() % (ARRAYSIZE(Threads) - 1) + 1);
+
+        for (uint32_t j = 0; j < Count; ++j) {
+            ASSERT_ON_FAILURE(CxPlatThreadCreate(&Config, &Threads[j]));
+        }
+
+        for (uint32_t j = 0; j < Count; ++j) {
+            CxPlatThreadWait(&Threads[j]);
+            CxPlatThreadDelete(&Threads[j]);
+        }
+    }
+
+    return 0;
+}
+
+#ifdef FUZZING
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
```

Comment:
How are you supposed to use this `data`?

Reply:
I'm planning
- GB.Buffers for StreamSend/DatagramSend
- GB.Alpns
- ResumptionTicket
- just cast to Server/Client/CredConfiguration
- SpinQuicSetRandomXXXParam

Reply:
I'm not convinced that's going to help much. Any "payload" buffer is just a pass through and doesn't really fuzz anything. That includes any send buffers and resumption tickets. You can't just cast random buffers to structs/handles and the code will definitely just crash (which is expected).

Are there other interfaces beyond `TestOneInput`? Ideally, I think having the fuzzer drive which API, one at a time might be better. Let's definitely talk more in today's Discord meeting.

Reply:
Created utility class which maintain each thread's pointer against data array from LibFuzzer.
And use it for API switching.
Fuzzer entry point is only LLVMFuzzerTestOneInput.
To use data more tricky is explained in this and its directory. Basically it is user to take care of data splitting
https://github.com/google/fuzzing/blob/master/docs/split-inputs.md